Sources/MatchingEngine/Registers.swift

```diff
@@ -10,7 +10,7 @@ extension Processor {
     var bools: Array<Bool> // TODO: bitset
 
     // currently, these are static readonly
-    var predicates: Array<(Element) -> Bool>
+    var predicates: Array<(Input, Input.Index) -> Input.Index?>
```

Comment:
Probably want to use a shared typealias, or maybe even a nominal type.

Reply:
Done.